src/App.tsx

```diff
@@ -1,26 +1,205 @@
-/* eslint-disable max-len */
+/* eslint-disable jsx-a11y/label-has-associated-control */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
 import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import {
+  USER_ID,
+  createTodo,
+  deleteTodo,
+  getTodos,
+  patchTodo,
+} from './api/todos';
+import { List } from './components/List';
+import { Error as ErrorMessage } from './components/Error';
+import { Footer } from './components/Footer';
+import { Header } from './components/Header';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState(Filter.All);
+  const [tempTodoTitle, setTempTodoTitle] = useState<string | null>('');
+  const ref = useRef<HTMLInputElement | null>(null);
+
+  const handleAddTodo = async (title: string) => {
+    const formattedTitle = title.trim();
+
+    if (!formattedTitle) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    const newTodo = {
+      title: formattedTitle,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    try {
+      setTempTodoTitle(title); // Show the temporary todo
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(currentTodos => [...currentTodos, createdTodo]);
+    } catch {
+      setErrorMessage('Unable to add a todo');
+      throw new Error('Unable to add a todo');
+    } finally {
+      setTempTodoTitle(null);
+    }
+  };
+
+  const handleDeleteTodo = async (id: number) => {
+    try {
+      await deleteTodo(id);
+
+      setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
+    } catch {
+      setErrorMessage('Unable to delete a todo');
+    }
+  };
+
+  const handleEditTodo = async (id: number, data: Partial<Todo>) => {
+    try {
+      const editedTodo = await patchTodo(id, data);
+
+      setTodos(currentTodos =>
+        currentTodos.map(todo => {
+          if (todo.id === id) {
+            return editedTodo;
+          }
+
+          return todo;
+        }),
+      );
+    } catch {
+      setErrorMessage('Unable to edit todo');
+    }
+  };
+
+  const clearCompletedTodos = async () => {
+    const filteredTodos = todos.filter(todo => todo.completed);
+
+    try {
+      const deleteCallback = async (todo: Todo) => {
+        try {
+          await deleteTodo(todo.id);
+
+          return { id: todo.id, status: 'resolved' };
```

Comment:
Do you really need to return these objects
I guess `Promise.allSettled` will do the same